### PR TITLE
Refactor zipArray to Throw Promise Error when Results Map has nil values

### LINKED ISSUE #74: Refactor zipArray to Throw Promise Error when Results Map has nil values
Instead of force unwrapping the result:
`let unwrappedResults = results.map { $0! }`

Detect the nil and throw:
`let unwrappedResults = results.map { $0 ?? Result { throw PromiseError.unknown } }`

### PR DESCRIPTION
**Pull request by mikerunnals**

### CODE REVIEW

File: Tests/PinkyPromiseTests/PromiseTest.swift [hunk 1]
```diff
@@ -857,5 +851,38 @@ class PromiseTest: XCTestCase {
...
+    
+    func callAndTestCompletion<T>(_ promise: Promise<T>, completion outerCompletion: @escaping (Result<T, Error>) -> Void) {
+        let completionCalled = expectation(description: "Completed")
+        promise.call { result in
+            outerCompletion(result)
+            completionCalled.fulfill()
+        }
+    }
+    
+    func testZipArray_ExposeForceUnwrapp() {
```
Comment (mikerunnals) [inline Tests/PinkyPromiseTests/PromiseTest.swift]: I'll rename this something good, like  testZipArrayThrowsUnknownErrorWhenFulfillCalledTwiceForOnePromiseAndZeroTimesFromSecondPromise

File: Sources/Promise.swift [hunk 2]
```diff
@@ -441,8 +441,13 @@ public func zipArray<T>(_ promises: [Promise<T>]) -> Promise<[T]> {
         }
         
         group.notify(queue: DispatchQueue.main) {
-            let unwrappedResults = results.map { $0! }
+            let unwrappedResults = results.map { $0 ?? Result { throw PromiseError.unknown } }
```
Comment (mikerunnals) [inline Sources/Promise.swift]: If the results here were enumerated instead of map, then the error thrown could include the index and even the promise that went unfulfilled. I wonder if this would help callers debug?

File: Sources/Promise.swift [hunk 3]
```diff
@@ -441,8 +441,15 @@ public func zipArray<T>(_ promises: [Promise<T>]) -> Promise<[T]> {
...
+            let unwrappedResults = results.enumerated().map { (index, result) in
+                result ?? Result { throw PromiseError.UnfufilledZipArrayPromise(index) }
+            }
             fulfill(zipArray(unwrappedResults))
         }
     }
 }
+
+public enum PromiseError : Error {
+    /// Index of Promise that was unfulfilled
```
Comment (mikerunnals) [inline Sources/Promise.swift]: Might add /// Index of **first** Promise that was unfulfilled

File: Sources/Promise.swift [hunk 4]
```diff
@@ -441,8 +441,15 @@ public func zipArray<T>(_ promises: [Promise<T>]) -> Promise<[T]> {
...
+                result ?? Result { throw PromiseError.UnfufilledZipArrayPromise(index) }
+            }
             fulfill(zipArray(unwrappedResults))
         }
     }
 }
+
+public enum PromiseError : Error {
+    /// Index of first unfulfilled Promise in zipArray
+    case UnfufilledZipArrayPromise(Int)
```
Comment (willtellis) [inline Sources/Promise.swift]: `unfufilledZipArrayPromise(Int)` is more conventional

File: Tests/PinkyPromiseTests/PromiseTest.swift [hunk 5]
```diff
@@ -857,5 +851,47 @@ class PromiseTest: XCTestCase {
...
+        promise.call { result in
+            outerCompletion(result)
+            completionCalled.fulfill()
+        }
+    }
+    
+    /// Test that zipArray throws Unknown Error when fulfill is called Twice for one Promise and Zero times for a second Promise
+    func testZipArrayThrowsUnknownErrorWhenUnwrappingResultsMapNil() {
+    
+        let overFilled: Promise<Int> = Promise<Int> { fulfill in
```
Comment (willtellis) [inline Tests/PinkyPromiseTests/PromiseTest.swift:900]: `Promise<Void>` might be better if the result values are unimportant to the tests

File: Tests/PinkyPromiseTests/PromiseTest.swift [hunk 6]
```diff
@@ -857,5 +851,47 @@ class PromiseTest: XCTestCase {
...
+        let unFilled: Promise<Int> = Promise<Int> { fulfill in
+        }
+        
+        callAndTestCompletion(zipArray([overFilled, filled, unFilled])) { result in
+            do {
+                _ = try result.get()
+                XCTFail("Expected to throw an error.")
+            } catch {
+                guard case PromiseError.UnfufilledZipArrayPromise(let index) = error else {
+                    XCTFail("Expected to throw an PromiseError.")
```
Comment (willtellis) [inline Tests/PinkyPromiseTests/PromiseTest.swift]: "_a_ PromiseError"

**Review by willtellis — COMMENTED** [review]
A few comments

File: Sources/Promise.swift [hunk 7]
```diff
@@ -441,8 +441,15 @@ public func zipArray<T>(_ promises: [Promise<T>]) -> Promise<[T]> {
...
+                result ?? Result { throw PromiseError.unfufilledZipArrayPromise(index) }
+            }
             fulfill(zipArray(unwrappedResults))
         }
     }
 }
+
+public enum PromiseError : Error {
+    /// Index of first unfulfilled Promise in zipArray
+    case unfufilledZipArrayPromise(Int)
```
Comment (willtellis) [inline Sources/Promise.swift]: Now that it's properly cased, I can see the spelling error: "unfulfilled" 😄

File: Tests/PinkyPromiseTests/PromiseTest.swift (hunk 5 above)
Comment (mikerunnals) [reply Tests/PinkyPromiseTests/PromiseTest.swift:900]: I tried it, but I didn't like it. `Promise<Void>` is kind of weird and edge-casey thing that might confuse a reader that the bug is only specific to special `Promise<Void>` promises.

File: Sources/Promise.swift (hunk 7 above)
Comment (mikerunnals) [reply Sources/Promise.swift]: lol, I think that's how I pronounce it 😬

File: Tests/PinkyPromiseTests/PromiseTest.swift (hunk 5 above)
Comment (willtellis) [reply Tests/PinkyPromiseTests/PromiseTest.swift:900]: Fair enough. Maybe a `String` with `"Test value"` as its value? Not a big deal either way

File: Sources/Promise.swift [hunk 8]
```diff
@@ -441,8 +441,15 @@ public func zipArray<T>(_ promises: [Promise<T>]) -> Promise<[T]> {
...
+                result ?? Result { throw PromiseError.unfulilledZipArrayPromise(index) }
+            }
             fulfill(zipArray(unwrappedResults))
         }
     }
 }
+
+public enum PromiseError : Error {
+    /// Index of first unfulfilled Promise in zipArray
+    case unfulilledZipArrayPromise(Int)
```
Comment (kconner) [inline Sources/Promise.swift]: I think you could make the API explain this case more clearly to a consumer:

```swift
case unfulfilledPromiseInZipArray(atIndex: Int)
```

**Comment by kconner** [comment]
The other `zip` functions in Promise.swift could produce this crash too: Instead of using an array of optionals, they use individual implicitly-unwrapped optionals to form a tuple, which is exactly as dangerous. Can we safeguard against them all and return a similar error for the tuple cases? Basically, eliminate all use of `!` in the library?

File: Sources/Promise.swift (hunk 8 above)
Comment (kconner) [reply Sources/Promise.swift]: Also, it's still misspelled, just differently misspelled :)

Comment (mikerunnals) [reply Sources/Promise.swift]: Google is telling me this: `unfulfilled Promise In Zip Array` is good (spelling wise)

**Comment by mikerunnals** [comment]
> The other `zip` functions in Promise.swift could produce this crash too: Instead of using an array of optionals, they use individual implicitly-unwrapped optionals to form a tuple, which is exactly as dangerous. Can we safeguard against them all and return a similar error for the tuple cases? Basically, eliminate all use of `!` in the library?

@kconner I'm happy to add this across the array functions. I'll re-post a PR with it tomorrow...

File: Sources/Promise.swift [hunk 9]
```diff
@@ -451,5 +451,5 @@ public func zipArray<T>(_ promises: [Promise<T>]) -> Promise<[T]> {
 
 public enum PromiseError : Error {
     /// Index of first unfulfilled Promise in zipArray
-    case unfulilledZipArrayPromise(Int)
+    case unfulilledZipArrayPromise(atIndex: Int)
```
Comment (kconner) [inline Sources/Promise.swift]: Bruh this still says "unfulilled"

File: Sources/Promise.swift (hunk 8 above)
Comment (kconner) [reply Sources/Promise.swift]: That search term's spelling does not match the code symbol, though

File: Sources/Promise.swift (hunk 9 above)
Comment (mikerunnals) [reply Sources/Promise.swift]: ah, I couldn't even see it! Thanks! Will fix...

File: Sources/Promise.swift [hunk 10]
```diff
@@ -441,8 +441,15 @@ public func zipArray<T>(_ promises: [Promise<T>]) -> Promise<[T]> {
...
+                result ?? Result { throw PromiseError.unfulfilledZipArrayPromise(atIndex: index) }
+            }
             fulfill(zipArray(unwrappedResults))
         }
     }
 }
+
+public enum PromiseError : Error {
+    /// Index of first unfulfilled Promise in zipArray
+    case unfulfilledZipArrayPromise(atIndex: Int)
```
Comment (kconner) [inline Sources/Promise.swift]: Now that we use this error in both `zip` and `zipArray`, can we drop the word "array" to make this error sound less specific?

```swift
case unfulfilledZipPromise(atIndex: Int)
```

or

```swift
case unfulfilledPromiseInZip(atIndex: Int)
```

File: Sources/Promise.swift [hunk 11]
```diff
@@ -393,10 +393,10 @@ public func zip<A, B, C, D>(_ promiseA: Promise<A>, _ promiseB: Promise<B>, _ pr
...
         let group = DispatchGroup()
 
-        var resultA: Result<A, Error>!
-        var resultB: Result<B, Error>!
-        var resultC: Result<C, Error>!
-        var resultD: Result<D, Error>!
+        var resultA = Result<A, Error> { throw PromiseError.unfulfilledZipArrayPromise(atIndex: 0) }
+        var resultB = Result<B, Error> { throw PromiseError.unfulfilledZipArrayPromise(atIndex: 1) }
+        var resultC = Result<C, Error> { throw PromiseError.unfulfilledZipArrayPromise(atIndex: 2) }
+        var resultD = Result<D, Error> { throw PromiseError.unfulfilledZipArrayPromise(atIndex: 3) }
```
Comment (kconner) [inline Sources/Promise.swift]: 👍

**Review by kconner — APPROVED** [review]
LGTM 💰
